enarx-keep-sgx/src/main.rs

```diff
@@ -16,65 +16,67 @@ use span::Span;
 
 use enclave::Leaf;
 
-fn main() {
+fn load() -> (enclave::Enclave, usize) {
     const USAGE: &str = "Usage: enarx-keep-sgx <shim> <code>";
 
-    let (enclave, entry) = {
-        // Get the arguments.
-        let mut args = std::env::args();
-        let shim = args.nth(1).expect(USAGE);
-        let code = args.next().expect(USAGE);
-
-        // Parse the shim and code and validate assumptions.
-        let shim = component::Component::from_path(shim).expect("Unable to parse shim");
-        let code = component::Component::from_path(code).expect("Unable to parse code");
-
-        // Determine the memory range for the enclave.
-        let span: Span<_, _> = shim.range().into();
-        let span = Span {
-            start: span.start,
-            count: span.count.next_power_of_two(),
-        };
-
-        // Initiate the enclave building process.
-        let mut builder = builder::Builder::new(span).expect("Unable to create builder");
-
-        // Load the shim segments.
-        for seg in shim.segments.iter() {
-            let mut src = unsafe { seg.src.align_to().1 };
-            let mut off = seg.dst.start - span.start;
-
-            // The first page of the shim entry is the TCS page.
-            if seg.dst.start == shim.entry {
-                builder
-                    .load(&src[..4096], off, SecInfo::tcs())
-                    .expect("Unable to add TCS page");
-                src = &src[4096..];
-                off += 4096;
-            }
-
-            if !src.is_empty() {
-                builder
-                    .load(src, off, SecInfo::reg(seg.rwx))
-                    .expect("Unable to add shim page");
-            }
+    // Get the arguments.
+    let mut args = std::env::args();
+    let shim = args.nth(1).expect(USAGE);
+    let code = args.next().expect(USAGE);
+
+    // Parse the shim and code and validate assumptions.
+    let shim = component::Component::from_path(shim).expect("Unable to parse shim");
+    let code = component::Component::from_path(code).expect("Unable to parse code");
+
+    // Determine the memory range for the enclave.
+    let span: Span<_, _> = shim.range().into();
+    let span = Span {
+        start: span.start,
+        count: span.count.next_power_of_two(),
+    };
+
+    // Initiate the enclave building process.
+    let mut builder = builder::Builder::new(span).expect("Unable to create builder");
+
+    // Load the shim segments.
+    for seg in shim.segments.iter() {
+        let mut src = unsafe { seg.src.align_to().1 };
+        let mut off = seg.dst.start - span.start;
+
+        // The first page of the shim entry is the TCS page.
+        if seg.dst.start == shim.entry {
+            builder
+                .load(&src[..4096], off, SecInfo::tcs())
+                .expect("Unable to add TCS page");
+            src = &src[4096..];
+            off += 4096;
```

Comment:
Should we pull in `memory` for `Page::size()` here to replace the `4096` literal?

Reply:
This happens already in future patches. This patch just moves existing code into a function.

Reply:
My point is that this is a good idea and I've already implemented it, but it is out of scope for this commit.

Reply:
Ah, I see. No worries then :slightly_smiling_face: